internal/decryption/decryption.go

```diff
@@ -90,19 +91,86 @@ func Decrypt(logReader io.Reader, out io.Writer, opts KeyProviderOpts) error {
 	return nil
 }
 
-func readAuditLogFile(logReader io.Reader) (AuditLogFormat, []string, error) {
-	const LineBreak = "\n"
+func peekFirstByte(reader io.ReadSeeker) (byte, error) {
+	b := make([]byte, 1)
+
+	n, err := reader.Read(b)
+	if err != nil {
+		return 0, err
+	}
+
+	if n != 1 {
+		return 0, fmt.Errorf("no bytes to read")
+	}
+
+	c, err := reader.Seek(0, io.SeekStart)
+	if err != nil {
+		return 0, err
+	}
+	if c != 0 {
+		return 0, fmt.Errorf("impossible to seek bytes")
+	}
+	return b[0], nil
+}
+
+func readAuditLogFile(reader io.ReadSeeker) (AuditLogFormat, []*AuditLogLine, error) {
 	auditLogFormat := BSON
 
-	data, err := io.ReadAll(logReader)
+	b, err := peekFirstByte(reader)
 	if err != nil {
 		return auditLogFormat, nil, err
 	}
 
-	const jsonStartChar = '{'
-	if len(data) > 0 && data[0] == jsonStartChar {
+	if b == '{' {
 		auditLogFormat = JSON
 	}
 
-	return auditLogFormat, strings.Split(string(data), LineBreak), nil
+	var logLines []*AuditLogLine
+
+	switch auditLogFormat {
+	case BSON:
+		logLines, err = readAuditLogFileBSON(reader)
+	case JSON:
+		logLines, err = readAuditLogFileJSON(reader)
+	}
+	return auditLogFormat, logLines, err
+}
+
+func readAuditLogFileBSON(reader io.ReadSeeker) ([]*AuditLogLine, error) {
+	var logLines []*AuditLogLine
+	for {
```

Comment:
the alternative would be like 
```
for raw, err := bson.NewFromIOReader(reader); err == nil || err != io.EOF {
    if err != nil {
       return nil, err
    }
    ....
}
```

I believe the break is still more readable, how about you?

Reply:
do we need to check err == nil?  this should do it, no? 
```
for raw, err := bson.NewFromIOReader(reader); err != io.EOF {
    if err != nil {
       return nil, err
    }
    ....
}
```

I don't have a strong preference, I just remember that I got a suggestion to change a for {} before. I'll leave my approval. Thanks for all the changes.

Reply:
I've investigated a bit further although we don't need `err == nil || ...` we still need `raw, err := bson.NewFromIOReader(reader)` on every loop

so something like `for raw, err := bson.NewFromIOReader(reader); err != io.EOF {` would never work because `raw, err := bson.NewFromIOReader(reader)` is only initialised once.

Ideally for this to work we would need to write something like `for ; raw, err := bson.NewFromIOReader(reader); err != io.EOF; {` unfortunately that does not work either because the evaluation section contain a `;` so then it moves the `err != io.EOF` to be the incrementor section of the for instead so no luck this time

Reply:
TIL. Thanks for the details Filipe :D